frame/support/src/traits/tokens/fungibles.rs

```diff
@@ -0,0 +1,142 @@
+	type AssetId: AssetId;
+	/// Scalar type for representing balance of an account.
+	type Balance: Balance;
+	/// The total amount of issuance in the system.
+	fn total_issuance(asset: Self::AssetId) -> Self::Balance;
+	/// The minimum balance any single account may have.
+	fn minimum_balance(asset: Self::AssetId) -> Self::Balance;
+	/// Get the `asset` balance of `who`.
+	fn balance(asset: Self::AssetId, who: &AccountId) -> Self::Balance;
+	/// Returns `true` if the `asset` balance of `who` may be increased by `amount`.
+	fn can_deposit(asset: Self::AssetId, who: &AccountId, amount: Self::Balance)
+		-> DepositConsequence;
+	/// Returns `Failed` if the `asset` balance of `who` may not be decreased by `amount`, otherwise
+	/// the consequence.
+	fn can_withdraw(
+		asset: Self::AssetId,
+		who: &AccountId,
+		amount: Self::Balance,
+	) -> WithdrawConsequence<Self::Balance>;
+}
+
+/// Trait for providing a set of named fungible assets which can be created and destroyed.
+pub trait Mutate<AccountId>: Inspect<AccountId> {
+	/// Attempt to increase the `asset` balance of `who` by `amount`.
+	///
+	/// If not possible then don't do anything. Possible reasons for failure include:
+	/// - Minimum balance not met.
+	/// - Account cannot be created (e.g. because there is no provider reference and/or the asset
+	///   isn't considered worth anything).
+	///
+	/// Since this is an operation which should be possible to take alone, if successful it will
+	/// increase the overall supply of the underlying token.
+	fn deposit(asset: Self::AssetId, who: &AccountId, amount: Self::Balance) -> DispatchResult;
+
+	/// Attempt to reduce the `asset` balance of `who` by `amount`.
+	///
+	/// If not possible then don't do anything. Possible reasons for failure include:
+	/// - Less funds in the account than `amount`
+	/// - Liquidity requirements (locks, reservations) prevent the funds from being removed
+	/// - Operation would require destroying the account and it is required to stay alive (e.g.
+	///   because it's providing a needed provider reference).
+	///
+	/// Since this is an operation which should be possible to take alone, if successful it will
+	/// reduce the overall supply of the underlying token.
+	///
+	/// Due to minimum balance requirements, it's possible that the amount withdrawn could be up to
+	/// `Self::minimum_balance() - 1` more than the `amount`. The total amount withdrawn is returned
+	/// in an `Ok` result. This may be safely ignored if you don't mind the overall supply reducing.
+	fn withdraw(asset: Self::AssetId, who: &AccountId, amount: Self::Balance)
+		-> Result<Self::Balance, DispatchError>;
+
+	/// Transfer funds from one account into another.
+	fn transfer(
+		asset: Self::AssetId,
+		source: &AccountId,
+		dest: &AccountId,
+		amount: Self::Balance,
+	) -> Result<Self::Balance, DispatchError> {
+		let extra = Self::can_withdraw(asset, &source, amount).into_result()?;
+		Self::can_deposit(asset, &dest, amount + extra).into_result()?;
```

Comment:
```suggestion
		Self::can_deposit(asset, &dest, amount + extra).into_result()?;
```

prefer saturating here